### PR TITLE
[2019.2.1] Use zeromq for multiprocessing logging

### PR DESCRIPTION
**Pull request by dwoz**
### What does this PR do?

Use zeromq for multiprocessing logging

### Tests written?
**[NOTICE] Bug fixes or features added to Salt require tests.**
Please review the [test documentation](https://docs.saltstack.com/en/latest/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite.

Yes

### Commits signed with GPG?

Yes

### CODE REVIEW

File: salt/log/setup.py [hunk 1]
```diff
@@ -463,6 +471,7 @@ def setup_temp_logger(log_level='error'):
     '''
     Setup the temporary console logger
     '''
+    return
```
Comment (s0undt3ch) [inline salt/log/setup.py]: You know what this one is for right? The temp logging...

File: salt/log/setup.py [hunk 2]
```diff
@@ -1088,6 +1259,69 @@ def patch_python_logging_handlers():
...
+        print("MP LOGGING TB %s" %('\n'.join(traceback.format_stack())))
+        raise
+    try:
+        while True:
+            try:
+                record = pickle.loads(puller.recv())
+                if record is None:
+                    # A sentinel to stop processing the queue
+                    break
+                # Just log everything, filtering will happen on the main process
```
Comment (s0undt3ch) [inline salt/log/setup.py]: This is where we should spend some time, if we know on the parent the desired log level, then all other log messages which don't meet the level requirements can be skipped going through _the wire_.

File: salt/log/handlers/__init__.py [hunk 3]
```diff
@@ -294,3 +294,60 @@ def enqueue(self, record):
...
+        record.exc_text = None
+        return record
+
+    def emit(self, record):
+        '''
+        Emit a record.
+
+        Writes the LogRecord to the queue, preparing it for pickling first.
+        '''
+        try:
```
[the remark this answers is not in the document]
Comment (dwoz) [reply salt/log/handlers/__init__.py]: pickle was just a quick and dirty way to get it working. I plan to look into msgpack still.

File: salt/log/setup.py (hunk 2 above)
Comment (dwoz) [reply salt/log/setup.py]: This should already be happening. I'm setting the log level on the ZMQHandler. https://github.com/saltstack/salt/pull/54276/files#diff-feeb9e1b18658ad0ea58ddd2d085bad7R992

File: salt/config/__init__.py [hunk 4]
```diff
@@ -1900,6 +1908,8 @@ def _gather_buffer_space():
     'pki_dir': os.path.join(salt.syspaths.CONFIG_DIR, 'pki', 'proxy'),
     'cachedir': os.path.join(salt.syspaths.CACHE_DIR, 'proxy'),
     'sock_dir': os.path.join(salt.syspaths.SOCK_DIR, 'proxy'),
+    'mp_logging_port': 3331,
+    'mp_logging_consumer': True,
 }
```
[the remark this answers is not in the document]
Comment (dwoz) [reply salt/config/__init__.py]: Right, still need to figure out some defaults for all the various processes. Right now I'm just pulling a random open port for the tests.

**Comment by dwoz** [comment]
w00t. got a passing build https://jenkinsci.saltstack.com/job/pr-kitchen-debian8-py2/job/PR-54276/5/

looks like there is some process cleanup issues though:
```
00:52:35         Error in atexit._run_exitfuncs:
00:52:35         Traceback (most recent call last):
00:52:35           File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
00:52:35             func(*targs, **kargs)
00:52:35           File "/usr/lib/python2.7/multiprocessing/util.py", line 321, in _exit_function
00:52:35             p._popen.terminate()
00:52:35           File "/usr/lib/python2.7/multiprocessing/forking.py", line 171, in terminate
00:52:35             os.kill(self.pid, signal.SIGTERM)
00:52:35         OSError: [Errno 3] No such process
00:52:35         Error in sys.exitfunc:
00:52:35         00:52:35,618 [tests.support.parser                                      :76  ][ERROR   ] An un-handled exception was caught by salt-testing's global exception handler:
00:52:35         OSError: [Errno 3] No such process
00:52:35         Traceback (most recent call last):
00:52:35           File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
00:52:35             func(*targs, **kargs)
00:52:35           File "/usr/lib/python2.7/multiprocessing/util.py", line 321, in _exit_function
00:52:35             p._popen.terminate()
00:52:35           File "/usr/lib/python2.7/multiprocessing/forking.py", line 171, in terminate
00:52:35             os.kill(self.pid, signal.SIGTERM)
00:52:35         OSError: [Errno 3] No such process
00:52:35         00:52:35,638 [salt.log.setup                                            :1460][ERROR   ] An un-handled exception was caught by salt's global exception handler:
00:52:35         OSError: [Errno 3] No such process
00:52:35         Traceback (most recent call last):
00:52:35           File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
00:52:35             func(*targs, **kargs)
00:52:35           File "/usr/lib/python2.7/multiprocessing/util.py", line 321, in _exit_function
00:52:35             p._popen.terminate()
00:52:35           File "/usr/lib/python2.7/multiprocessing/forking.py", line 171, in terminate
00:52:35             os.kill(self.pid, signal.SIGTERM)
00:52:35         OSError: [Errno 3] No such process
00:52:35         Traceback (most recent call last):
00:52:35           File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
00:52:35             func(*targs, **kargs)
00:52:35           File "/usr/lib/python2.7/multiprocessing/util.py", line 321, in _exit_function
00:52:35             p._popen.terminate()
00:52:35           File "/usr/lib/python2.7/multiprocessing/forking.py", line 171, in terminate
00:52:35             os.kill(self.pid, signal.SIGTERM)
00:52:35         OSError: [Errno 3] No such process
```

File: salt/config/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply salt/config/__init__.py]: :+1:

File: salt/log/handlers/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply salt/log/handlers/__init__.py]: :+1:

File: salt/log/setup.py (hunk 2 above)
Comment (s0undt3ch) [reply salt/log/setup.py]: Awesome.

File: salt/config/__init__.py [hunk 5]
```diff
@@ -371,6 +371,10 @@ def _gather_buffer_space():
     # The number of minutes between mine updates.
     'mine_interval': int,
 
+    'mp_logging_port': int,
+
+    'mp_logging_consumer': bool,
```
Comment (s0undt3ch) [inline salt/config/__init__.py]: I was going to ask why you added this, but a sip of coffee later I remembered, test suite logging consumer...

File: salt/log/handlers/__init__.py [hunk 6]
```diff
@@ -20,6 +20,13 @@
 from salt.log.mixins import NewStyleClassMixIn, ExcInfoOnLogLevelFormatMixIn
 from salt.ext.six.moves import queue
 
+import msgpack
+try:
+    import zmq
+    HAS_ZMQ = True
```
Comment (s0undt3ch) [inline salt/log/handlers/__init__.py]: We should really "blow up" if zmq is missing...
It's now a required piece to get logging working properly.
No?

File: salt/log/handlers/__init__.py [hunk 7]
```diff
@@ -162,135 +169,28 @@ class WatchedFileHandler(ExcInfoOnLogLevelFormatMixIn, logging.handlers.WatchedF
...
-            # bpo-35726: make copy of record to avoid affecting other handlers in the chain.
-            record = copy.copy(record)
-            record.message = msg
-            record.msg = msg
-            record.args = None
-            record.exc_info = None
-            record.exc_text = None
-            return record
+    def prepare(self, record):
+        return msgpack.dumps(record.__dict__, encoding='utf-8')
```
Comment (s0undt3ch) [inline salt/log/handlers/__init__.py]: We control both ends, use_bin_type?

File: salt/log/handlers/__init__.py [hunk 8]
```diff
@@ -162,135 +169,28 @@ class WatchedFileHandler(ExcInfoOnLogLevelFormatMixIn, logging.handlers.WatchedF
...
-                self.queue.put_nowait(record)
-            except queue.Full:
-                sys.stderr.write('[WARNING ] Message queue is full, '
-                                 'unable to write "{0}" to log'.format(record))
+        Writes the LogRecord to the queue, preparing it for pickling first.
+        '''
+        try:
+            self.sender.send(self.prepare(record))
+        except Exception:
+            self.handleError(record)
```
Comment (s0undt3ch) [inline salt/log/handlers/__init__.py]: You seem to have removed the queue handlers, 3rd-party code interacting with salt or using salt as a lib could depend on it.
Let's add it back?

File: salt/log/setup.py [hunk 9]
```diff
@@ -905,6 +952,63 @@ def setup_multiprocessing_logging_listener(opts, queue=None):
...
+        logging._acquireLock()  # pylint: disable=protected-access
+
+        if __MP_LOGGING_CONFIGURED is True:
+            return
+
+        # Let's set it to true as fast as possible
+        __MP_LOGGING_CONFIGURED = True
+
+        if os.getpid() in zmq_handlers:
+        #if __MP_LOGGING_ZMQ_HANDLER is not None:
```
Comment (s0undt3ch) [inline salt/log/setup.py]: Debugging leftover?
The same for `zmq_handlers`

File: salt/log/setup.py [hunk 10]
```diff
@@ -905,6 +952,63 @@ def setup_multiprocessing_logging_listener(opts, queue=None):
...
+        # exist, remove them.
+        __remove_null_logging_handler()
+        __remove_queue_logging_handler()
+
+        # Let's add a queue handler to the logging root handlers
+        #print("SET UP LOGGING %d %d" % ( os.getpid(), port or get_multiprocessing_logging_port()))
+        handler = ZMQHandler(port or get_multiprocessing_logging_port())
+#        zmq_handlers[os.getpid()] = handler
+        __MP_LOGGING_ZMQ_HANDLER = handler
+        logging.root.addHandler(handler)
```
Comment (s0undt3ch) [inline salt/log/setup.py]: The is probably what we did with the old MP logging, but in a child process we should probably clear any handlers and just have the zmq handler...

File: salt/log/setup.py [hunk 11]
```diff
@@ -1088,6 +1262,70 @@ def patch_python_logging_handlers():
         logging.handlers.QueueHandler = QueueHandler
 
 
+def __process_multiprocessing_logging_zmq(opts, port):
+    # Avoid circular import
+    import salt.utils.process
+    salt.utils.process.appendproctitle('MultiprocessingLoggingQueue')
```
Comment (s0undt3ch) [inline salt/log/setup.py]: `MultiprocessingLoggingZMQ`?

File: salt/log/setup.py [hunk 12]
```diff
@@ -1088,6 +1262,70 @@ def patch_python_logging_handlers():
...
+    puller = context.socket(zmq.PULL)
+    try:
+        puller.bind('tcp://127.0.0.1:{}'.format(port))
+    except Exception:
+        print("MP LOGGING TB %s" %('\n'.join(traceback.format_stack())))
+        raise
+    try:
+        while True:
+            try:
+                record_dict = msgpack.loads(puller.recv())
```
Comment (s0undt3ch) [inline salt/log/setup.py]: Why not the msgpack unpacker approach?
Because we always get full payloads?

File: salt/log/setup.py [hunk 13]
```diff
@@ -1183,6 +1421,7 @@ def __remove_temp_logging_handler():
     This function will run once logging has been configured. It just removes
     the temporary stream Handler from the logging handlers.
     '''
+    return
```
Comment (s0undt3ch) [inline salt/log/setup.py]: Leftover?

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: salt/log/handlers/__init__.py (hunk 7 above)
Comment (s0undt3ch) [reply salt/log/handlers/__init__.py]: This is going over the wire, you should also prepare the log record as the queue handler did.
There are object which won't travel over the wire, like traceback information, etc..

File: salt/log/handlers/__init__.py (hunk 6 above)
Comment (dwoz) [reply salt/log/handlers/__init__.py]: I had to add this for the dockermod and some ssh tests to pass.

File: salt/log/setup.py (hunk 12 above)
Comment (dwoz) [reply salt/log/setup.py]: Yes

File: salt/log/handlers/__init__.py (hunk 6 above)
Comment (s0undt3ch) [reply salt/log/handlers/__init__.py]: Ok

File: salt/log/setup.py (hunk 12 above)
Comment (s0undt3ch) [reply salt/log/setup.py]: Ok

File: salt/log/setup.py (hunk 13 above)
Comment (dwoz) [reply salt/log/setup.py]: yep, fixed

File: salt/log/setup.py (hunk 11 above)
Comment (dwoz) [reply salt/log/setup.py]: Fixed

File: salt/log/setup.py (hunk 9 above)
Comment (dwoz) [reply salt/log/setup.py]: yep, fixed

File: salt/log/handlers/__init__.py (hunk 8 above)
Comment (dwoz) [reply salt/log/handlers/__init__.py]: I added it back.

File: salt/minion.py [hunk 14]
```diff
@@ -1533,7 +1533,7 @@ def _handle_decoded_payload(self, data):
                 process.start()
         else:
             process.start()
-
+        #print("SUBPROC %d" % (process.pid))
```
Comment (dwoz) [inline salt/minion.py]: Fix this

File: salt/log/handlers/__init__.py (hunk 7 above)
Comment (dwoz) [reply salt/log/handlers/__init__.py]: I changed the prepare function to prepare the log record like the queue handler and also added the use_bin_type flag for msgpack.

File: salt/log/setup.py (hunk 10 above)
Comment (dwoz) [reply salt/log/setup.py]: I've fixed this.

File: salt/log/setup.py [hunk 15]
```diff
@@ -98,6 +107,7 @@
     'process': TextFormat('bold', 'blue'),
 }
 
+zmq_handlers = {}
```
Comment (s0undt3ch) [inline salt/log/setup.py]: Leftover?

File: salt/log/setup.py [hunk 16]
```diff
@@ -905,6 +972,61 @@ def setup_multiprocessing_logging_listener(opts, queue=None):
...
+        # Since noone will process them, memory usage will grow. If they
+        # exist, remove them.
+        __remove_null_logging_handler()
+        __remove_queue_logging_handler()
+
+        # Let's add a queue handler to the logging root handlers
+        handler = ZMQHandler(port or get_multiprocessing_logging_port())
+        __MP_LOGGING_ZMQ_HANDLER = handler
+        if __MP_IN_MAINPROCESS is True:
+            for handler in logging.root.handlers:
```
Comment (s0undt3ch) [inline salt/log/setup.py]: You overwrite the previously defined `handler` like this.

File: salt/log/setup.py [hunk 17]
```diff
@@ -995,6 +1117,30 @@ def shutdown_temp_logging():
...
+    global __MP_LOGGING_ZMQ_HANDLER
+
+    if not __MP_LOGGING_CONFIGURED or not __MP_LOGGING_ZMQ_HANDLER:
+        return
+
+    try:
+        #handler = zmq_handlers.pop(os.getpid())
+        logging._acquireLock()
+        # Let's remove the queue handler from the logging root handlers
+        logging.root.removeHandler(__MP_LOGGING_ZMQ_HANDLER)
```
Comment (s0undt3ch) [inline salt/log/setup.py]: Shouldn't you also `.stop()` the handler?

File: salt/transport/zeromq.py [hunk 18]
```diff
@@ -844,14 +844,15 @@ def __init__(self, opts):
     def connect(self):
         return tornado.gen.sleep(5)
 
-    def _publish_daemon(self, log_queue=None):
+    def _publish_daemon(self, *args, **kwargs):
```
Comment (s0undt3ch) [inline salt/transport/zeromq.py]: Do we really have to use args and kwargs here?

File: tests/integration/logging/test_zmq_logging.py [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+    def __init__(self, level=logging.NOTSET):
+        super(TestLogHandler, self).__init__(level=level)
+        self.messages = []
+
+    def handle(self, record):
+        self.messages.append(record)
+
+
+def test_target(host, port):
```
Comment (s0undt3ch) [inline tests/integration/logging/test_zmq_logging.py]: If this isn't a test don't name it like `test_*`, once we get this into neon and under pytest we'd have to rename it.

File: tests/integration/logging/test_zmq_logging.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+        self.logger.addHandler(self.handler)
+        # Run the consumer in a thread so we can access the handler's messages
+        # attribute later.
+        self.log_consumer = threading.Thread(
+            target=salt.log.setup._process_multiprocessing_logging_zmq,
+            args=(opts, self.port, True)
+        )
+        self.log_consumer.start()
+        # Allow time for the consumer to get all connected
+        time.sleep(5)
```
Comment (s0undt3ch) [inline tests/integration/logging/test_zmq_logging.py]: So we need this much for it to connect?

File: tests/integration/logging/test_zmq_logging.py [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+
+    def send_consumer_shutdown(self):
+        context = zmq.Context()
+        sender = context.socket(zmq.PUSH)
+        sender.connect('tcp://{}:{}'.format(self.host, self.port))
+        time.sleep(1)
+        try:
+            sender.send(msgpack.dumps(None))
+        finally:
+            time.sleep(5)
```
Comment (s0undt3ch) [inline tests/integration/logging/test_zmq_logging.py]: And to terminate?

File: tests/integration/netapi/test_client.py [hunk 22]
```diff
@@ -26,6 +27,8 @@ def setUp(self):
         Set up a NetapiClient instance
         '''
         opts = salt.config.client_config(os.path.join(TMP_CONF_DIR, 'master'))
+        #import pprint
```
Comment (s0undt3ch) [inline tests/integration/netapi/test_client.py]: Leftover

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Pull request closed without merging**